src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -289,36 +292,115 @@ float LightTree::evaluate_node_pdf(
+    }
+}
+
 float LightTree::compute_node_probability(
     const LightTreeNode<AABB3d>&    node,
     const AABB3d&                   bbox,
-    const Vector3d&                 surface_point) const
+    const ShadingPoint&             shading_point) const
 {
-    // Calculate probability of a single node based on its distance
-    // to the surface point being illuminated.
-    // For leaf nodes use the actual position of the light source, instead
-    // of center of the bbox. It is more precise for shaped lights.
-    Vector3d position;
+    // Calculate probability of a single node based on its contribution over solid angle.
+    const float r = bbox.radius();
+    const float r2 = r * r;
+    const float rcp_surface_area = 1.0f / r2;
+
+    // Triangle centroid is a more precise pocision than the center of the bbox.
     const Item& item = m_items[node.get_item_index()];
-    if (node.is_leaf() && item.m_light_type == EmittingTriangleType)
-    {
-        const size_t light_index = item.m_light_index;
-        const EmittingTriangle& triangle = m_emitting_triangles[light_index];
-        // Use centroid as triangle position approximation.
-        position = (triangle.m_v0 + triangle.m_v1 + triangle.m_v2) * (1.0 / 3.0);
-    }
-    else
-        position = bbox.center();
+    const Vector3d position = (node.is_leaf() && item.m_light_type == EmittingTriangleType)
+        ? emitting_triangle_centroid(item.m_light_index)
+        : bbox.center();
+
+    const Vector3d surface_point = shading_point.get_point();
 
     const float squared_distance =
         static_cast<float>(square_distance(surface_point, position));
 
-    return node.get_importance() / squared_distance;
+    // Evaluated point is outside the bbox.
+    // The original Nathan's implementation returns importance divided by the node surface area.
+    // However, replacing the surface area by squared distance showed to result in less noise.
+    if (squared_distance <= r2)
+        return node.get_importance() / squared_distance;
+    //
+    // Implementation of Lambertian lighting model for sub-hemispherical light sources.
+    // Reference:
+    //  [1] Area Light Sources for Real-Time Graphics
+    //      https://www.microsoft.com/en-us/research/wp-content/uploads/1996/03/arealights.pdf
+    //
+    const Vector3d outcoming_light_direction = normalize(bbox.center() - surface_point);
+    const float sin_theta2 = min(1.0f, (r2 / squared_distance));
+    const float cos_theta = sqrt(1.0f - sin_theta2);
+
+    const Vector3d incoming_light_direction = shading_point.get_ray().m_dir;
+
+    // [1] "Arbitrary direction D receives light only if dot(D,L) >= 0".
+    const Vector3d N = (dot(shading_point.get_geometric_normal(), incoming_light_direction) <= 0.0f)
+        ? normalize(shading_point.get_shading_normal())
+        : -normalize(shading_point.get_shading_normal());
+
+    const float cos_omega = min(1.0, max(-1.0, dot(N, outcoming_light_direction)));
```

Comment:
You can also use `foundation::clamp()`:

    const float cos_ometa = clamp(dot(N, outcoming_light_direction), -1.0f, 1.0f);

Note that you must use the `f` suffix for float literals, otherwise it won't compile in VS.